File: scripts/render-agents.ts
#!/usr/bin/env node

import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative } from "node:path";
import { fileURLToPath } from "node:url";

import { readToml } from "../lib/toml.ts";
import type { TomlTable, TomlValue } from "../lib/toml.ts";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));

function readText(path: string): string {
  return new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
}

function valueAsTable(value: TomlValue | undefined): TomlTable {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? value
    : Object.create(null) as TomlTable;
}

function valueAsStrings(value: TomlValue | undefined): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

function tomlString(value: string): string {
  return JSON.stringify(value);
}

function tomlMultiline(value: string): string {
  const escaped = value.replaceAll("\\", "\\\\").replaceAll('"""', '\\"\\"\\"');
  return `"""${escaped.trimEnd()}\\n"""`;
}

export function renderCodex(name: string, spec: TomlTable, prompt: string): string {
  const provider = valueAsTable(spec.codex);
  const description = spec.description;
  if (typeof description !== "string") throw new TypeError("role description must be a string");
  const lines = [`name = ${tomlString(name)}`, `description = ${tomlString(description)}`];
  const candidates = valueAsStrings(spec.nickname_candidates);
  if (candidates.length > 0) {
    lines.push(`nickname_candidates = [${candidates.map(tomlString).join(", ")}]`);
  }
  if (provider.model) lines.push(`model = ${tomlString(String(provider.model))}`);
  if (provider.reasoning_effort) {
    lines.push(`model_reasoning_effort = ${tomlString(String(provider.reasoning_effort))}`);
  }
  lines.push(`developer_instructions = ${tomlMultiline(prompt)}`);
  return `${lines.join("\n")}\n`;
}

function yamlScalar(value: string): string {
  return JSON.stringify(value);
}

export function renderClaude(name: string, spec: TomlTable, prompt: string): string {
  const provider = valueAsTable(spec.claude);
  const description = spec.description;
  if (typeof description !== "string") throw new TypeError("role description must be a string");
  const lines = ["---", `name: ${name}`, `description: ${yamlScalar(description)}`];
  if (provider.model) lines.push(`model: ${String(provider.model)}`);
  if (provider.permission_mode) lines.push(`permissionMode: ${String(provider.permission_mode)}`);
  const tools = valueAsStrings(provider.tools);
  if (tools.length > 0) lines.push(`tools: ${tools.join(", ")}`);
  if (provider.max_turns) lines.push(`maxTurns: ${String(provider.max_turns)}`);
  const skills = valueAsStrings(spec.skills);
  if (skills.length > 0) {
    lines.push("skills:");
    lines.push(...skills.map((skill) => `  - ${skill}`));
  }
  if (provider.color) lines.push(`color: ${String(provider.color)}`);
  lines.push("---", "", prompt.trimEnd(), "");
  return lines.join("\n");
}

function bytewiseCompare(left: string, right: string): number {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function generatedFiles(base: string): string[] {
  if (!existsSync(base)) return [];
  return readdirSync(base)
    .filter((name) => !name.startsWith("."))
    .map((name) => join(base, name))
    .filter((path) => statSync(path).isFile())
    .sort(bytewiseCompare);
}

function usage(stream: NodeJS.WriteStream): void {
  stream.write("usage: render-agents.ts [-h] [--check]\n");
}

function main(): number {
  const args = process.argv.slice(2);
  if (args.some((argument) => argument === "-h" || argument === "--help")) {
    if (args.length !== 1) {
      usage(process.stderr);
      return 2;
    }
    usage(process.stdout);
    return 0;
  }
  if (args.some((argument) => argument !== "--check") || args.filter((argument) => argument === "--check").length > 1) {
    usage(process.stderr);
    return 2;
  }
  const check = args.includes("--check");
  const outputs = new Map<string, string>();
  const rolesRoot = join(ROOT, "roles");
  for (const entry of readdirSync(rolesRoot).sort(bytewiseCompare)) {
    const roleDirectory = join(rolesRoot, entry);
    if (!statSync(roleDirectory).isDirectory()) continue;
    const spec = readToml(join(roleDirectory, "role.toml"));
    const prompt = readText(join(roleDirectory, "ROLE.md"));
    const name = spec.name;
    if (typeof name !== "string") throw new TypeError(`role name must be a string: ${entry}`);
    const codex = valueAsTable(spec.codex);
    const claude = valueAsTable(spec.claude);
    if (codex.enabled !== false) {
      outputs.set(join(ROOT, ".codex", "agents", `${name}.toml`), renderCodex(name, spec, prompt));
    }
    if (claude.enabled !== false) {
      outputs.set(join(ROOT, ".claude", "agents", `${name}.md`), renderClaude(name, spec, prompt));
    }
  }

  const generatedRoots = [join(ROOT, ".codex", "agents"), join(ROOT, ".claude", "agents")];
  const stale = generatedRoots.flatMap(generatedFiles).filter((path) => !outputs.has(path));
  const changed = [...stale];
  for (const [path, content] of outputs) {
    const current = existsSync(path) ? readText(path) : undefined;
    if (current !== content) {
      changed.push(path);
      if (!check) {
        mkdirSync(dirname(path), { recursive: true });
        writeFileSync(path, content, "utf8");
        console.log(`wrote ${relative(ROOT, path)}`);
      }
    }
  }

  if (!check) {
    for (const path of stale) {
      unlinkSync(path);
      console.log(`removed ${relative(ROOT, path)}`);
    }
  }

  if (check && changed.length > 0) {
    console.log("generated agent definitions are stale:");
    for (const path of changed) console.log(`  ${relative(ROOT, path)}`);
    return 1;
  }
  return 0;
}

try {
  process.exitCode = main();
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exitCode = 1;
}

File: scripts/render-agents.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import tomllib
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def toml_string(value: str) -> str:
    return json.dumps(value, ensure_ascii=False)


def toml_multiline(value: str) -> str:
    escaped = value.replace('\\', '\\\\').replace('"""', '\\"\\"\\"')
    return f'"""{escaped.rstrip()}\\n"""'


def render_codex(name: str, spec: dict, prompt: str) -> str:
    provider = spec.get("codex", {})
    lines = [
        f'name = {toml_string(name)}',
        f'description = {toml_string(spec["description"])}',
    ]
    candidates = spec.get("nickname_candidates", [])
    if candidates:
        lines.append("nickname_candidates = [" + ", ".join(toml_string(x) for x in candidates) + "]")
    if provider.get("model"):
        lines.append(f'model = {toml_string(provider["model"])}')
    if provider.get("reasoning_effort"):
        lines.append(f'model_reasoning_effort = {toml_string(provider["reasoning_effort"])}')
    lines.append(f'developer_instructions = {toml_multiline(prompt)}')
    return "\n".join(lines) + "\n"


def yaml_scalar(value: str) -> str:
    return json.dumps(value, ensure_ascii=False)


def render_claude(name: str, spec: dict, prompt: str) -> str:
    provider = spec.get("claude", {})
    lines = ["---", f"name: {name}", f"description: {yaml_scalar(spec['description'])}"]
    if provider.get("model"):
        lines.append(f"model: {provider['model']}")
    if provider.get("permission_mode"):
        lines.append(f"permissionMode: {provider['permission_mode']}")
    tools = provider.get("tools", [])
    if tools:
        lines.append("tools: " + ", ".join(tools))
    if provider.get("max_turns"):
        lines.append(f"maxTurns: {provider['max_turns']}")
    skills = spec.get("skills", [])
    if skills:
        lines.append("skills:")
        lines.extend(f"  - {skill}" for skill in skills)
    if provider.get("color"):
        lines.append(f"color: {provider['color']}")
    lines.extend(["---", "", prompt.rstrip(), ""])
    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(description="Render provider-specific agent definitions from canonical Hanchou roles")
    parser.add_argument("--check", action="store_true", help="fail if generated files differ")
    args = parser.parse_args()

    outputs: dict[Path, str] = {}
    for role_dir in sorted((ROOT / "roles").iterdir()):
        if not role_dir.is_dir():
            continue
        spec = tomllib.loads((role_dir / "role.toml").read_text())
        prompt = (role_dir / "ROLE.md").read_text()
        name = spec["name"]
        if spec.get("codex", {}).get("enabled", True):
            outputs[ROOT / ".codex" / "agents" / f"{name}.toml"] = render_codex(name, spec, prompt)
        if spec.get("claude", {}).get("enabled", True):
            outputs[ROOT / ".claude" / "agents" / f"{name}.md"] = render_claude(name, spec, prompt)

    generated_roots = (ROOT / ".codex" / "agents", ROOT / ".claude" / "agents")
    stale = [path for base in generated_roots for path in base.glob("*") if path.is_file() and path not in outputs]
    changed = list(stale)
    for path, content in outputs.items():
        current = path.read_text() if path.exists() else None
        if current != content:
            changed.append(path)
            if not args.check:
                path.parent.mkdir(parents=True, exist_ok=True)
                path.write_text(content)
                print(f"wrote {path.relative_to(ROOT)}")

    if not args.check:
        for path in stale:
            path.unlink()
            print(f"removed {path.relative_to(ROOT)}")

    if args.check and changed:
        print("generated agent definitions are stale:")
        for path in changed:
            print(f"  {path.relative_to(ROOT)}")
        raise SystemExit(1)


if __name__ == "__main__":
    main()
